src/main/java/net/rubygrapefruit/platform/internal/jni/AbstractFileEventFunctions.java

```diff
@@ -46,25 +50,47 @@ public AbstractWatcherBuilder(FileWatcherCallback callback) {
      * Configures a new watcher using a builder. Call {@link AbstractWatcherBuilder#start()} to
      * actually start the {@link FileWatcher}.
      */
-    public abstract AbstractWatcherBuilder newWatcher(FileWatcherCallback callback);
+    public abstract AbstractWatcherBuilder newWatcher(BlockingQueue<FileWatchEvent> queue);
 
     protected static class NativeFileWatcherCallback {
-        private final FileWatcherCallback delegate;
 
-        public NativeFileWatcherCallback(FileWatcherCallback delegate) {
-            this.delegate = delegate;
+        private final BlockingQueue<FileWatchEvent> eventQueue;
+
+        public NativeFileWatcherCallback(BlockingQueue<FileWatchEvent> eventQueue) {
+            this.eventQueue = eventQueue;
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void pathChanged(int type, String path) {
-            delegate.pathChanged(FileWatcherCallback.Type.values()[type], path);
+        public void pathChanged(int typeIndex, String path) throws InterruptedException {
+            FileWatchEvent.Type type = FileWatchEvent.Type.values()[typeIndex];
+            if (type == FileWatchEvent.Type.OVERFLOWED) {
+                signalOverflow(path);
+            } else {
+                queueEvent(new ChangeEvent(type, path), false);
+            }
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void reportError(Throwable ex) {
-            delegate.reportError(ex);
+        public void reportError(Throwable ex) throws InterruptedException {
+            queueEvent(new ErrorEvent(ex), true);
+        }
+
+        private void queueEvent(FileWatchEvent event, boolean deliverOnOverflow) throws InterruptedException {
+            // TODO Make the timeout configurable
+            if (!eventQueue.offer(event, 1, SECONDS)) {
+                NativeLogger.LOGGER.info("Event queue overflow, dropping all events");
+                signalOverflow(null);
+                if (deliverOnOverflow) {
+                    eventQueue.put(event);
```

Comment:
Let's not block here, but fail if the queue is unable to accept the error event.

Reply:
We aren't blocking anymore.